android/sdl_android/src/main/java/com/smartdevicelink/managers/lockscreen/LockScreenDeviceIconManager.java

```diff
@@ -0,0 +1,164 @@
+
+        File f = new File(this.context.getCacheDir() + "/" + STORED_ICON_PATH, iconHash);
+        ByteArrayOutputStream bos = new ByteArrayOutputStream();
+        icon.compress(Bitmap.CompressFormat.PNG, 0 /*ignored for PNG*/, bos);
+        byte[] bitmapdata = bos.toByteArray();
+
+        FileOutputStream fos = null;
+        try {
+            fos = new FileOutputStream(f);
+            fos.write(bitmapdata);
+            fos.flush();
+            fos.close();
+        } catch (Exception e) {
+            DebugTool.logError("Failed to save icon to cache");
+            e.printStackTrace();
+            return;
+        }
+
+        JSONObject iconParams;
+        try {
+            iconParams = buildDeviceIconParameters(f.getAbsolutePath());
+            writeDeviceIconParametersToSystemPreferences(iconHash, iconParams);
+        } catch (JSONException e) {
+            DebugTool.logError("Failed to save to shared preferences, clearing cache icon directory");
+            clearIconDirectory();
+            e.printStackTrace();
+        }
+    }
+
+    Bitmap getFileFromCache(String iconUrl) {
+        String iconHash = getMD5HashFromIconUrl(iconUrl);
+        SharedPreferences sharedPref = this.context.getSharedPreferences(SDL_DEVICE_STATUS_SHARED_PREFS, Context.MODE_PRIVATE);
+        String iconParameters = sharedPref.getString(iconHash, null);
+
+        if (iconParameters != null) {
+            JSONObject jsonObject = null;
+            try {
+                jsonObject = new JSONObject(iconParameters);
+                String storedUrl = jsonObject.getString(STORED_URL);
+                Bitmap cachedIcon = BitmapFactory.decodeFile(storedUrl);
+                if(cachedIcon == null) {
+                    DebugTool.logError("Failed to get Bitmap from decoding file cache");
+                    clearIconDirectory();
+                    return null;
+                } else {
+                    return cachedIcon;
+                }
+            } catch (JSONException e) {
+                DebugTool.logError("Failed to get file from cache, removing shared pref");
+                sharedPref.edit().remove(iconHash).commit();
+                e.printStackTrace();
+                return null;
+            }
+        } else {
+            DebugTool.logError("Failed to get system preferences");
+            return null;
+        }
+    }
+
+    private void writeDeviceIconParametersToSystemPreferences(String iconHash, JSONObject jsonObject) {
+        SharedPreferences sharedPref = this.context.getSharedPreferences(SDL_DEVICE_STATUS_SHARED_PREFS, Context.MODE_PRIVATE);
+        SharedPreferences.Editor editor = sharedPref.edit();
+        editor.putString(iconHash, jsonObject.toString());
+        editor.commit();
+    }
+
+    private JSONObject buildDeviceIconParameters(String storedUrl) throws JSONException {
+        JSONObject parametersJson = new JSONObject();
+        parametersJson.put(STORED_URL, storedUrl);
```

Comment:
If I understand it correctly, this saves the file path, not the url right? If so, can we change the name to `filePath`  to make it easier to read?

Reply:
Refactored